client.d.ts

```diff
@@ -0,0 +1,316 @@
+   * If wrapData is true it will allow you to send all the messages to only one topic by wrapping the siofu actions and messages.
+   *
+   * ```js
+   * {
+   *   action: 'complete',
+   *   message: {
+   *     id: id,
+   *     success: success,
+   *     detail: fileInfo.clientDetail
+   *   }
+   * }
+   * ```
+   *
+   * If wrapData is an object constituted of two mandatory key and one optional:
+   * wrapKey and unwrapKey (mandatory): Corresponding to the key used to wrap the siofu data and message
+   * additionalData (optional): Corresponding to the data to send along with file data
+   *
+   * ```js
+   * if wrapData = {
+   *   wrapKey: {
+   *     action: 'actionType',
+   *     message: 'data'
+   *   },
+   *   unwrapKey: {
+   *     action: 'actionType',
+   *     message: 'message'
+   *   },
+   *   additionalData: {
+   *     acknowledgement: true
+   *   }
+   * }
+   * ```
+   *
+   * When Siofu will send for example a complete message this will send:
+   * ```js
+   * {
+   *   acknowledgement: true,
+   *   actionType: 'complete',
+   *   data: {
+   *     id: id,
+   *     success: success,
+   *     detail: fileInfo.clientDetail
+   *   }
+   * }
+   * ```
+   *
+   * and it's waiting from client data formatted like this:
+   * ```js
+   * {
+   *   actionType: '...',
+   *   message: {...}
+   * }
+   * ```
+   * @description
+   * /!\ If wrapData is wrong configured is interpreted as false /!\
+   *
+   * @default false
+   */
+  wrapData: boolean
+
+	/**
+	 * Allow user to access to some private function to customize message reception.
+	 * This is used if you specified wrapOptions on the client side and have to manually bind message to callback.
+   * @default false
+	 */
+  exposePrivateFunction: boolean
+}
+
+export type SocketIOFileUploadClientChoose =
+{
+  files: FileList
+}
+
+export type SocketIOFileUploadClientStart<Metadata> =
+{
+  file: File & {
+    meta: Metadata
+  }
+}
+
+export type SocketIOFileUploadClientProgress<Metadata> =
+{
+  file: SocketIOFileUploadClientStart<Metadata>['file']
+
+  /**
+   * The number of bytes that have been loaded into memory.
+   */
+  bytesLoaded: number
+
+  /**
+   * The filename to which the server saved the file.
+   */
+  name: string
+}
+
+export type SocketIOFileUploadClientLoad<Metadata> =
+{
+  file: SocketIOFileUploadClientStart<Metadata>['file']
+
+  /**
+   * An instance of a W3C FileReader object.
+   */
+  reader: FileReader
+
+  /**
+   * The filename to which the server saved the file.
+   */
+  name: string
+}
+
+export type SocketIOFileUploadClientComplete<Metadata> =
+{
+  file: SocketIOFileUploadClientStart<Metadata>['file']
+
+  /**
+   * `true` if the server-side implementation ran without error; `false` otherwise.
+   */
+  success: boolean
+
+  /**
+   * The value of file.clientDetail on the server side. Properties may be added to this object literal during any event on the server side.
+   */
+  detail: Metadata
```

Comment:
Note that the `clientDetail` and the `meta` can be different types.

I'm okay with making these generic but I'm also okay if you make them both `any`

Reply:
Okay, each class (Client/Server) now accepts two generics: `Metadata` and `ClientDetail`, which serve this purpose.
https://github.com/sffc/socketio-file-upload/pull/144/commits/625f700c6c4b1ee0c1f91d1978c4c0e728358375